### PR TITLE
fix: Apply redactions to room state events in database, too

### LINKED ISSUE #890: Redacting a room state event does not appear to redact the event in the local store
**Describe the bug**
I'm redacting a state event with `JoinedRoom::redact(state_event_id)` and the trace log shows that the following sync contains the redaction event. When I retrieve the original state event after that with `JoinedRoom::get_state_event_static(state_event_id)` I get the original event (enum variant), not the redacted one. @jplatte mentioned that it should already update the local store and redact the event there, similar to what synapse does. An alternative might be to re-fetch the redacted event from synapse.

**To Reproduce**
1. Send state event
2. Redact with `JoinedRoom::redact(state_event_id)`
3. Wait for sync
4. Fetch state event again via `JoinedRoom::get_state_event_static(state_event_id)`

**Expected behavior**
The state event should be the redacted enum variant of the event.

### PR DESCRIPTION
**Pull request by gnunicorn**
fixes #890

### CODE REVIEW

**Comment by poljar** [comment]
Am I looking at the wrong place or are the new tests not run?

image

File: testing/matrix-sdk-integration-testing/src/tests/redaction.rs [hunk 1]
```diff
@@ -34,7 +34,14 @@ async fn test_redaction() -> Result<()> {
 
     room.send_state_event(content, "").await?;
     // sync up.
-    tamatoa.sync_once(Default::default()).await?;
+    for _ in 0..2 {
+        // we call sync up to three times to give the server time to flush other
```
Comment (jplatte) [inline testing/matrix-sdk-integration-testing/src/tests/redaction.rs:38]: `0..2` has two elements: 0, and 1.

Comment (Hywan) [reply testing/matrix-sdk-integration-testing/src/tests/redaction.rs:38]: And `0..=2` has three elements, 0, 1 and 2.

File: testing/matrix-sdk-integration-testing/src/tests/redaction.rs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    // let's send a specific state event
+    let avatar_url = mxc_uri!("mxc://example.org/avatar").to_owned();
+    let content = assign!(RoomAvatarEventContent::new(), {
+        url: Some(avatar_url),
+    });
+
+    room.send_state_event(content, "").await?;
+    // sync up.
+    for _ in 0..2 {
+        // we call sync up to three times to give the server time to flush other
```
Comment (Hywan) [inline testing/matrix-sdk-integration-testing/src/tests/redaction.rs:93]: Same.

File: testing/matrix-sdk-integration-testing/src/tests/redaction.rs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        tamatoa.sync_once(Default::default()).await?;
+        if room.get_state_event_static::<RoomAvatarEventContent>("").await?.is_some() {
+            break;
+        }
+    }
+
+    // check state event.
+
+    let room_avatar_event: SyncRoomAvatarEvent =
+        room.get_state_event_static("").await?.expect("Room Avatar not found").deserialize()?;
```
Comment (Hywan) [inline testing/matrix-sdk-integration-testing/src/tests/redaction.rs:104]: Passing an empty string seems like the `get_state_event_static` method should take a `Option<&str>` or something similar, thoughts?

Comment (gnunicorn) [reply testing/matrix-sdk-integration-testing/src/tests/redaction.rs:104]: Not sure... the spec states that `state_key` is required, but may be an empty string for certain types. And as `"" != None`, I wouldn't agree that this was a better API (as under that format `None` could also be understood as "don't match the specific state_key, just hand out the first that matches the type").

Comment (jplatte) [reply testing/matrix-sdk-integration-testing/src/tests/redaction.rs:104]: We have strong state key types in Ruma. I don't think `Option` makes sense, but we could have separate `get_state_event_static` + `get_state_event_static_with_key` functions where the former works for (most) state events which have [`StateKey = EmptyStateKey`](https://docs.ruma.io/ruma_common/events/room/name/struct.RoomNameEventContent.html#impl-StateEventContent-for-RoomNameEventContent) and the latter works for everything and has stronger types (so getting a member event would require you to pass [a User ID](https://docs.ruma.io/ruma_common/events/room/member/struct.RoomMemberEventContent.html#impl-StateEventContent-for-RoomMemberEventContent) for example).

File: testing/matrix-sdk-integration-testing/src/tests/redaction.rs [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        Some(token) => SyncSettings::default().token(token),
+        None => SyncSettings::default(),
+    };
+    let bg_syncer = tokio::spawn(async move { bg_sync.sync(settings).await });
+
+    let room = tamatoa.create_room(request).await?;
+    bg_syncer.abort();
+    // let's send a specific state event
+
+    let content = RoomNameEventContent::new(Some("Inapropriate text".to_owned()));
```
Comment (jplatte) [inline testing/matrix-sdk-integration-testing/src/tests/redaction.rs:47]:
```suggestion
    let content = RoomNameEventContent::new(Some("Inappropriate text".to_owned()));
```

File: crates/matrix-sdk-base/src/store/memory_store.rs [hunk 5]
```diff
@@ -351,10 +352,18 @@ impl MemoryStore {
             }
         }
 
+        let make_room_version = |room_id| {
+            self.room_info
+                .get(room_id)
+                .and_then(|info| info.room_version().cloned())
+                .unwrap_or_else(|| {
+                    warn!(%room_id, "Unable to find the room version, assuming version 9");
```
Comment (jplatte) [inline crates/matrix-sdk-base/src/store/memory_store.rs:360]: Why would we default to v9?

File: crates/matrix-sdk-base/src/store/memory_store.rs [hunk 6]
```diff
@@ -461,6 +461,27 @@ impl MemoryStore {
             }
         }
 
+        for (room_id, redactions) in &changes.redactions {
+            let mut room_version = None;
+            if let Some(room) = self.room_state.get_mut(room_id) {
+                for mut ref_room_mu in room.iter_mut() {
+                    for mut ref_evt_mu in ref_room_mu.value_mut().iter_mut() {
```
Comment (jplatte) [inline crates/matrix-sdk-base/src/store/memory_store.rs:468]: I don't think these two `&mut`s need to have the extra `mut` of the binding.

File: testing/matrix-sdk-integration-testing/src/helpers.rs [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        .with(tracing_subscriber::EnvFilter::from_default_env())
+        .with(tracing_subscriber::fmt::layer().with_test_writer())
+        .init();
+}
+
+/// read the test configuration from the environment
+pub fn test_server_conf() -> (String, String) {
+    (
+        option_env!("HOMSERVER_URL").unwrap_or("http://localhost:8228").to_owned(),
+        option_env!("HOMSERVER_DOMAIN").unwrap_or("matrix-sdk.rs").to_owned(),
```
Comment (jplatte) [inline testing/matrix-sdk-integration-testing/src/helpers.rs:29]:
```suggestion
        option_env!("HOMESERVER_URL").unwrap_or("http://localhost:8228").to_owned(),
        option_env!("HOMESERVER_DOMAIN").unwrap_or("matrix-sdk.rs").to_owned(),
```

File: crates/matrix-sdk-base/src/store/memory_store.rs (hunk 5 above)
Comment (gnunicorn) [reply crates/matrix-sdk-base/src/store/memory_store.rs:360]: this is just a code move from below. I don't know if that is a correct assumption ...

File: crates/matrix-sdk-base/src/store/memory_store.rs (hunk 6 above)
Comment (gnunicorn) [reply crates/matrix-sdk-base/src/store/memory_store.rs:468]: Unfortunately, they do:
```
   Compiling matrix-sdk-base v0.5.1 (/home/ben/Dev/matrix/matrix-rust-sdk/crates/matrix-sdk-base)
error[E0596]: cannot borrow `ref_room_mu` as mutable, as it is not declared as mutable
   --> crates/matrix-sdk-base/src/store/memory_store.rs:470:39
    |
469 |                 for ref_room_mu in room.iter_mut() {
    |                     ----------- help: consider changing this to be mutable: `mut ref_room_mu`
470 |                     for ref_evt_mu in ref_room_mu.value_mut().iter_mut() {
    |                                       ^^^^^^^^^^^^^^^^^^^^^^^ cannot borrow as mutable

error[E0596]: cannot borrow `ref_evt_mu` as mutable, as it is not declared as mutable
   --> crates/matrix-sdk-base/src/store/memory_store.rs:471:39
    |
470 |                     for ref_evt_mu in ref_room_mu.value_mut().iter_mut() {
    |                         ---------- help: consider changing this to be mutable: `mut ref_evt_mu`
471 |                         let raw_evt = ref_evt_mu.value_mut();
    |                                       ^^^^^^^^^^^^^^^^^^^^^^ cannot borrow as mutable

   [ .... ]
   Compiling matrix-sdk-base v0.5.1 (/home/ben/Dev/matrix/matrix-rust-sdk/crates/matrix-sdk-base)
error[E0596]: cannot borrow `ref_room_mu` as mutable, as it is not declared as mutable
   --> crates/matrix-sdk-base/src/store/memory_store.rs:470:43
    |
469 |                 for ref_room_mu in room.iter_mut() {
    |                     ----------- help: consider changing this to be mutable: `mut ref_room_mu`
470 |                     for mut ref_evt_mu in ref_room_mu.value_mut().iter_mut() {
    |                                           ^^^^^^^^^^^^^^^^^^^^^^^ cannot borrow as mutable

For more information about this error, try `rustc --explain E0596`.
error: could not compile `matrix-sdk-base` due to previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `matrix-sdk-base` due to previous error
```

File: crates/matrix-sdk-base/src/store/memory_store.rs (hunk 5 above)
Comment (jplatte) [reply crates/matrix-sdk-base/src/store/memory_store.rs:360]: Okay, can leave it as-is then, I guess.

File: crates/matrix-sdk-base/src/store/memory_store.rs (hunk 6 above)
Comment (jplatte) [reply crates/matrix-sdk-base/src/store/memory_store.rs:470]: Ah, `DashMap::iter_mut()` doesn't return `&mut T`, it returns a `RefMutMulti` type which has this `fn value_mut(&mut self)`. Okay then.

Comment (gnunicorn) [reply crates/matrix-sdk-base/src/store/memory_store.rs:470]: yeah, it's a bit weird of an API, but I think it is about pushing the write-lock until the last possible moment ... but it leads to this sort of ugliness in the using syntax...

**Review by jplatte — APPROVED** [review]

**Pull request merged**